chore: Improve the current tutorial for consuming orchestration and grounding (#1)
shibeshduw/Tutorials: tutorials/ai-core-orchestration-consumption-opt/ai-core-orchestration-consumption-opt.md

- Here, we extend the use case introduced in Previous Tutorial, where orchestration was executed without incorporating data masking or content filtering. Here, we will include these advanced modules to improve data privacy, security, and response quality. 

**NOTE** : If you are continuing with the same project from the previous tutorial, skip steps 1 and 2. Otherwise, create a new project using the already deployed orchestration URL to access the Harmonized API.

> For detailed installation and usage of the **SAP Cloud SDK for AI (JavaScript)**, visit the official [GitHub repository](https://github.com/SAP/ai-sdk-js/tree/main?tab=readme-ov-file#sap-ai-sdkorchestration). This page provides comprehensive steps to set up, integrate and test the SDK effectively in your projects.

Reviewer: [req] Can we align this part with the below `ai-api` section? And pointing out `@sap-ai-sdk/orchestration` would be helpful to the user IMO.

[OPTION BEGIN [JavaScript SDK ]]

The next step involves creating a template that specifies how the AI should handle the CV content. The template will include both `SystemMessage` and `UserMessage` components. 

• `SystemMessage`: Defines the AI assistant's role and instructions. 

> • `UserMessage`: Represents the user's input to be processed. 

Reviewer: [req] I think `SystemMessage` and `UserMessage` are python specific classes or? We should not use these terms.

```javascript

// Define the system and user messages 

> const templatingConfig = { 

Reviewer: [req] tl;dr always type config. In the upcoming version of the orchestration spec, due to the introduction of string literal for roles, any config without a type (or eventually string cannot be assigned to string literal) will cause an issue.

```javascript

import { buildAzureContentSafetyFilter } from '@sap-ai-sdk/orchestration'; 

const inputFilter = buildAzureContentSafetyFilter({
  Hate: 'ALLOW_ALL',
  SelfHarm: 'ALLOW_SAFE',
  Sexual: 'ALLOW_SAFE_LOW_MEDIUM',
  Violence: 'ALLOW_SAFE_LOW_MEDIUM'
});

const outputFilter = buildAzureContentSafetyFilter({
  Hate: 'ALLOW_ALL',
  SelfHarm: 'ALLOW_SAFE',
  Sexual: 'ALLOW_SAFE_LOW_MEDIUM',
  Violence: 'ALLOW_SAFE_LOW_MEDIUM'
});

> const filteringConfig = {

Reviewer: [req] Same here, although not a big problem here, but type it would be preferred.

Reviewer: And import then also the type.

> const filteringConfig = {

Reviewer: [pp] Would be nice if we can add a hint here that this config will be used later in which section.

  filtering: {
    input: { filters: [inputFilter] },
    output: { filters: [outputFilter] }
  }
}

> console.log('Content Filtering configuration defined successfully.');

Reviewer: [pp] I would remove this console log from the previous version. Looks really useless to me.

```

**NOTE** : Adjust thresholds for hate, sexual, self-harm, and violence categories based on your use case. 

> Multiple filters can be applied for both input and output. In this tutorial, we are using Azure Content Filter, but you can choose from the available providers based on your use case. For more information on using this and/or additional modules, please refer to the official documentation of [SAP Cloud SDK for AI (Javascript)](https://github.com/SAP/ai-sdk-js/tree/main/packages/orchestration).

Reviewer: ```suggestion
Multiple content filters can be applied for both input and output. In this tutorial, we use Azure Content Safety filter, but you can choose from the available providers based on your use case. For more information, please refer to the official documentation of [`@sap-ai-sdk/orchestration`](https://github.com/SAP/ai-sdk-js/tree/main/packages/orchestration) package.
```

[OPTION BEGIN [JavaScript SDK]]

> **Generate Responses for Multiple Models** 

Reviewer: ```suggestion
**Generate Responses with Multiple Models** 
```

This step outlines the process of generating responses for a set of queries using different models. The `generateResponsesForModels()` function iterates through each model and executes queries with the created template. 

**Note**: Ensure that your orchestration deployment is in Running Status and ready to be consumed during this process.

```javascript

import { writeFile } from 'fs/promises';
import { OrchestrationClient } from '@sap-ai-sdk/orchestration'; 

const RESOURCE_GROUP = "YourResourceGroupId"; // Define the resource group, change this to your resource group name

// Generate responses from multiple models using OrchestrationClient

> async function generateResponsesForModels(cvContent) { 

Reviewer: [req] type `cvContent`

> async function generateResponsesForModels(cvContent) { 

Reviewer: [pp] Here I also don't like the naming `generateResponsesForModels`, could it be `generateResponsesWIthModels`?

  // Initialize OrchestrationClient asynchronously for list of models
    const responses = await Promise.all(
      models.map(async (model) => {
        const orchestrationClient = new OrchestrationClient(
          {
            llm: {
              model_name: model,
              model_params: { 
                max_tokens: 1000, 
                temperature: 0.6, 
              }, 
            },
            ...templatingConfig,

>             ...dataMaskingConfig,

Reviewer: [q] I assume data masking is not part of this tutorial right? But it is weird to me that Java also has this. Can you double check?

Author: It is a part of the same tutorial. `generateResponsesWithModels()` uses both filtering and masking to generate responses. It helps to look at the [tutorial](https://developers.sap.com/tutorials/ai-core-orchestration-consumption-opt.html) and the changes to see how the tutorial flows.

            ...filteringConfig
          },
          { resourceGroup: RESOURCE_GROUP } 
        );

        try { 
          // Run orchestration with the provided input (candidate resume content) 
          const response = await orchestrationClient.chatCompletion({ 
            inputParams: { candidate_resume: cvContent }, 
          }); 

          // Extract the response content and return it
          return { 
            model, 
            response: response.getContent(), 
          }; 
        } catch (error: any) { 
          const errorDetails = {

>             status: error.cause?.status || 500,

Reviewer: [pp] console error a 500 from SDK feels really weird. In our sample code, it is an express server, and it makes sense there to send a 500 response. Can we restructure the error handling and make it easier to understand? Or maybe link our [Error Handling](https://github.com/SAP/ai-sdk-js/tree/main?tab=readme-ov-file#error-handling) doc here.

            message: error.cause?.response?.data || error.message,
          };
          console.error(`Error with model ${model}:`, errorDetails);
        } 
      })
    );

    // Optionally save the responses to a file
    await writeFile( 
      'model_responses_js.txt', 
      responses 
        .map((res) => `Response from model ${res.model}:\n${res.response}\n${'-'.repeat(80)}\n`) 
        .join(''), 
      'utf-8' 
    ); 
} 

// Example usage
const modelResponses = await generateResponsesForModels(cvContent); 
console.log(`\n=== Responses for model: ${modelResponses.model} ===\n`);

> console.log(modelResponses.response);

Reviewer: [req] What does `generateResponsesForModels()` function return? And I assume `modelResponses` should be accessed via some index?

Author: Ah yes good point. I would need an index because `modelResponses` is a list. I have mapped the response accordingly.
